src/lib/homestore/shard_manager.cpp

```diff
@@ -1,15 +1,263 @@
 #include "homeobject.hpp"
+#include "replication_message.hpp"
+
+#include <homestore/homestore.hpp>
+#include <homestore/blkdata_service.hpp>
+#include <homestore/meta_service.hpp>
 
 namespace homeobject {
 
 uint64_t ShardManager::max_shard_size() { return Gi; }
+uint64_t ShardManager::max_shard_num_in_pg() {return ((uint64_t)0x01) << 48;}
+
+shard_id HSHomeObject::generate_new_shard_id(pg_id pg) {
+    std::scoped_lock lock_guard(_pg_lock);
+    auto iter = _pg_map.find(pg);
+    RELEASE_ASSERT(iter != _pg_map.end(), "Missing pg info");
+    auto new_sequence_num = ++(iter->second.shard_sequence_num);
+    RELEASE_ASSERT(new_sequence_num < ShardManager::max_shard_num_in_pg(), "new shard id must be less than ShardManager::max_shard_num_in_pg()");
+    return make_new_shard_id(pg, new_sequence_num);
+}
+
+shard_id HSHomeObject::make_new_shard_id(pg_id pg, uint64_t sequence_num) const {
+    return ((uint64_t)pg << 48) | sequence_num; 
+}
+
+uint64_t HSHomeObject::get_sequence_num_from_shard_id(uint64_t shard_id) {
+    return shard_id & (0x0000FFFFFFFFFFFF);
```

Comment:
Should we do `& (max_shard_num_in_pg() - 1)` rather than a raw bitmask?

Reply:
good point, will update that.